File: Python/백준/Silver/1931. 회의실 배정/회의실 배정.py
import sys
N = int(sys.stdin.readline())
me = {}
etc = {}
mt = 0
for _ in range(N):
    a, b = map(int, sys.stdin.readline().split())
    mt = max(mt, b)
    if a == b:
        try:
            me[b] += 1
        except:
            me[b] = 1
    else:
        try: etc[b] = max(etc[b], a)
        except: etc[b] = a
        
end = 0
cnt = 0
for i in range(mt + 1):
    if i < end: continue
    #print(cnt ,i)
    tmp = cnt
    if i in me: cnt += me[i]
    if i in etc: cnt += int(etc[i] >= end)
    if tmp != cnt: end = i
print(cnt)

File: Java/백준/Silver/1931. 회의실 배정/회의실 배정.java
import java.io.*;
import java.util.*;
class Meet {
    public int s;
    public int e;
    Meet(int s, int e) {
        this.s = s;
        this.e = e;
    }

    public String toString() {
        return "(" + this.s + ", " + this.e + ")";
    }
}
class Solution {
    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

    public void solution() throws IOException{
        int N = Integer.parseInt(br.readLine());
        ArrayList<Meet> al = new ArrayList<>();
        for (int i = 0; i < N; i++) {
            String tStr[] = br.readLine().split(" ");
            int s = Integer.parseInt(tStr[0]);
            int e = Integer.parseInt(tStr[1]);
            al.add(new Meet(s, e));
        }
        Collections.sort(al, new Comparator<Meet>() {
            @Override
            public int compare(Meet a, Meet b) {
                if (a.e < b.e) return -1;
                else if (a.e > b.e) return 1;
                else {
                    if (a.s < b.s) return -1;
                    else return 1;
                }

            }
        });
        int last = al.get(0).e;
        int cnt = 1;
        for (int i = 1; i < N; i++) {
            Meet now = al.get(i);
            if (last <= now.s) {
                last = now.e;
                cnt += 1;
            }
        }
        bw.write(cnt + "\n");
        bw.flush();
        bw.close();
    }
}

public class Main {
    public static void main(String argv[]) throws IOException{
        Solution sol = new Solution();
        sol.solution();
    }
}